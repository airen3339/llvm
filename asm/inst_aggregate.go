package asm

import (
	"fmt"

	"github.com/llir/llvm/ir"
	"github.com/llir/ll/ast"
)

// --- [ Aggregate instructions ] ----------------------------------------------

// ~~~ [ extractvalue ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstExtractValue(inst ir.Instruction, old *ast.ExtractValueInst) (*ir.InstExtractValue, error) {
	i, ok := inst.(*ir.InstExtractValue)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstExtractValue, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ insertvalue ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstInsertValue(inst ir.Instruction, old *ast.InsertValueInst) (*ir.InstInsertValue, error) {
	i, ok := inst.(*ir.InstInsertValue)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstInsertValue, got %T", inst))
	}
	// TODO: implement
	return i, nil
}
