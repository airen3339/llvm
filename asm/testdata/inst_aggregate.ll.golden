define i32 @extractvalue_1() {
; <label>:0
	%result = extractvalue { i8, i32 } { i8 21, i32 42 }, 1
	ret i32 %result
}

define i32 @extractvalue_2() {
; <label>:0
	%result = extractvalue { i32, { [2 x i32], i8 } } { i32 0, { [2 x i32], i8 } { [2 x i32] [i32 100, i32 42], i8 11 } }, 1, 0, 1
	ret i32 %result
}

define i32 @extractvalue_3() {
; <label>:0
	%result = extractvalue { i8, i32 } { i8 21, i32 42 }, 1, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define { i32, i32 } @insertvalue_1() {
; <label>:0
	%result = insertvalue { i32, i32 } { i32 21, i32 42 }, i32 42, 0
	ret { i32, i32 } %result
}

define { i32, { [2 x i32], i32 } } @insertvalue_2() {
; <label>:0
	%result = insertvalue { i32, { [2 x i32], i32 } } { i32 42, { [2 x i32], i32 } { [2 x i32] [i32 100, i32 42], i32 42 } }, i32 42, 1, 0, 0
	ret { i32, { [2 x i32], i32 } } %result
}

define { i32, i32 } @insertvalue_3() {
; <label>:0
	%result = insertvalue { i32, i32 } { i32 21, i32 42 }, i32 42, 0, !baz !{!"qux"}, !foo !{!"bar"}
	ret { i32, i32 } %result
}
